pkg/metrics/shoot_customization.go

```diff
@@ -0,0 +1,643 @@
+			if !ok {
+				return nil, nil, utils.NewTypeConversionError()
+			}
+
+			var extensionCounter = map[string]float64{}
+			for _, s := range shoots {
+				if s.Spec.Extensions != nil {
+					for _, extension := range s.Spec.Extensions {
+						extensionCounter[extension.Type]++
+					}
+				}
+			}
+			values, labels := mapLabelAndValues(&extensionCounter)
+			return values, labels, nil
+		},
+	},
+
+	// Kube API Server customization.
+	{
+		Name:   fmt.Sprintf("%s_apiserver_basicauth_total", metricShootsCustomPrefix),
+		Help:   "Count of Shoots which have basic auth enabled on the kube apiserver.",
+		Labels: []string{},
+		Type:   template.Gauge,
+		CollectFunc: func(obj interface{}, params ...interface{}) (*[]float64, *[][]string, error) {
+			shoots, ok := obj.([]*gardenv1alpha1.Shoot)
+			if !ok {
+				return nil, nil, utils.NewTypeConversionError()
+			}
+			var counter = make([]float64, 1, 1)
+			for _, s := range shoots {
+				if s.Spec.Kubernetes.KubeAPIServer != nil && s.Spec.Kubernetes.KubeAPIServer.EnableBasicAuthentication != nil && *s.Spec.Kubernetes.KubeAPIServer.EnableBasicAuthentication {
+					counter[0]++
+				}
+			}
+			return &counter, &[][]string{}, nil
+		},
+	},
+	{
+		Name:   fmt.Sprintf("%s_apiserver_auditpolicy_total", metricShootsCustomPrefix),
+		Help:   "Count of Shoots which have an audit log policy configured for the kube apiserver.",
+		Labels: []string{},
+		Type:   template.Gauge,
+		CollectFunc: func(obj interface{}, params ...interface{}) (*[]float64, *[][]string, error) {
+			shoots, ok := obj.([]*gardenv1alpha1.Shoot)
+			if !ok {
+				return nil, nil, utils.NewTypeConversionError()
+			}
+			var counter = make([]float64, 1, 1)
+			for _, s := range shoots {
+				if s.Spec.Kubernetes.KubeAPIServer != nil && s.Spec.Kubernetes.KubeAPIServer.AuditConfig != nil && s.Spec.Kubernetes.KubeAPIServer.AuditConfig.AuditPolicy != nil && s.Spec.Kubernetes.KubeAPIServer.AuditConfig.AuditPolicy.ConfigMapRef != nil {
+					counter[0]++
+				}
+			}
+			return &counter, &[][]string{}, nil
+		},
+	},
+	{
+		Name:   fmt.Sprintf("%s_apiserver_oidcconfig_total", metricShootsCustomPrefix),
+		Help:   "Count of Shoots which have an open id connect coniguration for the kube apiserver.",
+		Labels: []string{},
+		Type:   template.Gauge,
+		CollectFunc: func(obj interface{}, params ...interface{}) (*[]float64, *[][]string, error) {
+			shoots, ok := obj.([]*gardenv1alpha1.Shoot)
+			if !ok {
+				return nil, nil, utils.NewTypeConversionError()
+			}
+			var counter = make([]float64, 1, 1)
+			for _, s := range shoots {
+				if s.Spec.Kubernetes.KubeAPIServer != nil && s.Spec.Kubernetes.KubeAPIServer.OIDCConfig != nil {
+					counter[0]++
+				}
+			}
+			return &counter, &[][]string{}, nil
+		},
+	},
+	{
+		Name:   fmt.Sprintf("%s_apiserver_featuregates_total", metricShootsCustomPrefix),
+		Help:   "Count of Shoots with enabled kube apiserver feature gates.",
+		Labels: []string{"featuregate"},
```

Comment:
I do not see any metrics with the label `featuregate`. Is this expected behavior?

Reply:
Maybe you tested it with a Gardener which does not manage Shoots with apiserver feature gate configuration.
```
$ curl localhost:2718/metrics | grep feature
# HELP garden_shoots_custom_apiserver_featuregates_total Count of Shoots with enabled kube apiserver feature gates.
# TYPE garden_shoots_custom_apiserver_featuregates_total gauge
garden_shoots_custom_apiserver_featuregates_total{featuregate="CustomResourceWebhookConversion"} 53
garden_shoots_custom_apiserver_featuregates_total{featuregate="ExpandInUsePersistentVolumes"} 1
garden_shoots_custom_apiserver_featuregates_total{featuregate="TTLAfterFinished"} 13
garden_shoots_custom_apiserver_featuregates_total{featuregate="VolumeSnapshotDataSource"} 1
```

Reply:
Ok I see the metrics now. Should we initialize the metrics somehow? If there are no featuregates set then `garden_shoots_custom_apiserver_featuregates_total` will be missing.

Reply:
Since we need the featuregate label I don't think we can easily initialize this metric.